trace/trace_test.go

```diff
@@ -582,3 +582,63 @@ func TestBucket(t *testing.T) {
 		}
 	}
 }
+
+type exporter []*SpanData
+func (e *exporter) Export(s *SpanData) {
+	*e = append(*e, s)
+}
+
+func Test_Issue328_EndSpanTwice(t *testing.T) {
+	var e exporter
+	RegisterExporter(&e)
+	defer UnregisterExporter(&e)
+	ctx := context.Background()
+	ctx = StartSpanWithOptions(ctx, "span-1", StartSpanOptions{Sampler: AlwaysSample()})
+	EndSpan(ctx)
+	EndSpan(ctx)
+	UnregisterExporter(&e)
+	if len(e) != 1 {
+		t.Fatalf("expected only a single span, got %#v", e)
+	}
+}
+
+func TestStartSpanAfterEnd(t *testing.T) {
+	var spans exporter
+	RegisterExporter(&spans)
+	defer UnregisterExporter(&spans)
+	ctx := StartSpanWithOptions(context.Background(), "parent", StartSpanOptions{Sampler: AlwaysSample()})
+	ctx1 := StartSpanWithOptions(ctx, "span-1", StartSpanOptions{Sampler: AlwaysSample()})
+	EndSpan(ctx1)
+	// Start a new span with the context containing span-1
+	// even though span-1 is ended, we still add this as a new child of span-1
+	ctx2 := StartSpanWithOptions(ctx1, "span-2", StartSpanOptions{Sampler: AlwaysSample()})
+	EndSpan(ctx2)
+	EndSpan(ctx)
+	UnregisterExporter(&spans)
+	if len(spans) != 3 {
+		t.Fatalf("expected 3 spans, got %#v", spans)
+	}
+	var span1, span2, parent *SpanData
+	for _, span := range spans {
+		switch span.Name {
+		case "span-1":
+			span1 = span
+			break
+		case "span-2":
+			span2 = span
+			break
+		case "parent":
+			parent = span
+			break
+		}
+	}
+	if span1.TraceID != parent.TraceID || span2.TraceID != parent.TraceID {
+		t.Errorf("expected same TraceID, got: %q, %q, %q", parent.TraceID, span1.TraceID, span2.TraceID)
+	}
+	if span1.ParentSpanID != parent.SpanID {
+		t.Errorf("span-1 parent, expected: %q got: %q", parent.SpanID, span1.ParentSpanID)
```

Comment:
Use got, want terminology. Got first, want second.

You can omit got if you already have an identifier for what's got in the message:

"span-1 parent = %v; want %v"